ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,351 @@
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  public void registerHeartbeat(HostInfo hostInfo, long timestamp) {
+    receivedHeartbeats.putIfAbsent(hostInfo, new TreeMap<>());
+    TreeMap<Long, HeartbeatInfo> heartbeats = receivedHeartbeats.get(hostInfo);
+
+    synchronized (heartbeats) {
+      heartbeats.put(timestamp, new HeartbeatInfo(timestamp));
+    }
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
+          missedCount = (ts - prev - 1) / configs.heartbeatSendIntervalMs;
+        } else {
+          missedCount = 0;
+        }
+        prev = ts;
+      }
+      // Check frame from last received heartbeat to window end
+      if (windowEnd - prev - 1 > 0) {
```

Comment:
sorry don't follow this well.. :/  could you explain?

Reply:
In order to count the missed heartbeats, we check the elapsed time between received heartbeats. We might have received a heartbeat at the beginning of the window and then nothing. So we cannot use the elapsed time between heartbeats but we use the window end as the barrier.

Reply:
Okay fair.. I still have some gaps in understanding... The if- check here, seems to check if the difference between the last heartbeat (which can be windowStart or the real last heartbeat) and the windowEnd is non-zero.. I think this will mostly pass, in normal mode of operation right?  There will be some non-zero gap between heartbeat and windowEnd, as I can imagine.. 

The assignment below, resets the `missedCount` (presumably since we are interested in consecutive misses only).. I am wondering if thats the right thing to do..  could it be possible that we already accumulated a missedCount in the for loop above and we could reset this to 0? if say heartbeatSendIntervalMs was 10ms and heartbeats arrive perfectly (let say), then windowEnd-prev is say 10ms and we set `missedCount = 10-1/10` as 0? even though we detected enough consecutive missed above?